diff --git a/Cargo.lock b/Cargo.lock
index 2f5ad19..73031fd 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -91,6 +91,37 @@ version = "1.0.102"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "7f202df86484c868dbad7eaa557ef785d5c66295e41b460ef922eca0723b842c"
 
+[[package]]
+name = "argh"
+version = "0.1.18"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "16a3bf61b195d0d2202b21a5783b273ccba8dc04c7f07c3bc22a70d02b81e154"
+dependencies = [
+ "argh_derive",
+ "argh_shared",
+]
+
+[[package]]
+name = "argh_derive"
+version = "0.1.18"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "8ba478f46c1c645512959fada39ced2b0d3552fb4a60ef3ff65d0c8850d222a5"
+dependencies = [
+ "argh_shared",
+ "proc-macro2",
+ "quote",
+ "syn",
+]
+
+[[package]]
+name = "argh_shared"
+version = "0.1.18"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "2db5bbb2053ca6b6d64f11e8572f2e4a0a32421e87f4260ec49039cf75b07ef3"
+dependencies = [
+ "serde",
+]
+
 [[package]]
 name = "arrayref"
 version = "0.3.9"
@@ -4196,6 +4227,7 @@ name = "xr-chaperone"
 version = "0.1.0"
 dependencies = [
  "anyhow",
+ "argh",
  "ash",
  "bytemuck",
  "glam 0.32.1",
diff --git a/Cargo.toml b/Cargo.toml
index 23f44e4..62d364b 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -40,9 +40,10 @@ anyhow = "1"
 
 # Shared States
 parking_lot = "0.12.5"
+argh = "0.1.18"
 
 [build-dependencies]
 shaderc = "0.10.1"
 
 [profile.release]
-opt-level = 3
\ No newline at end of file
+opt-level = 3
diff --git a/src/main.rs b/src/main.rs
index bee7beb..dbac43f 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -17,10 +17,18 @@ use config::Config;
 use std::sync::Arc;
 use tracing::{error, info};
 use xdg::BaseDirectories;
+use argh::FromArgs;
 
 const APP_NAME: &str = env!("CARGO_PKG_NAME");
 const CONFIG_FILE: &str = "chaperone.toml";
 
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh( description = "xr-chaperone")]
+pub struct CliArgs {
+    #[argh(switch, short = 's', description = "run without a GUI for use in a systemd service, don't use this if you haven't configured xr-chaperone with the GUI at least once.")]
+    service_mode: bool,
+}
+
 /// Entry point, loads the config, starts the xr thread, spawns the interface
 fn main() -> Result<()> {
     tracing_subscriber::fmt()
@@ -29,6 +37,7 @@ fn main() -> Result<()> {
         )
         .init();
 
+    let cli_args: CliArgs = argh::from_env();
     let xdg_dirs = BaseDirectories::with_prefix(APP_NAME);
     let cfg_path = xdg_dirs.place_config_file(CONFIG_FILE);
     if let Err(e) = cfg_path {
@@ -58,7 +67,7 @@ fn main() -> Result<()> {
     // Spawn XR render thread
     let xr_state = Arc::clone(&state);
     let xr_cfg = cfg.clone();
-    std::thread::Builder::new()
+    let xr_thread_handle = std::thread::Builder::new()
         .name("xr-render".into())
         .spawn(move || xr_thread::run_xr_thread(xr_state, xr_cfg))?;
 
@@ -82,7 +91,12 @@ fn main() -> Result<()> {
                 error!("Failed to Set Offsets: {}", e);
             }
 
-            ui::run(state, cfg, cfg_path)?;
+            if cli_args.service_mode == true {
+                info!("Requested service mode, not starting the GUI.");
+                let res = xr_thread_handle.join();
+            } else {
+                ui::run(state, cfg, cfg_path)?;
+            }
         }
     }
 
