diff --git a/build_scripts/qt/compilers/gcc.pri b/build_scripts/qt/compilers/gcc.pri
index dc86954..a2ad4d9 100644
--- a/build_scripts/qt/compilers/gcc.pri
+++ b/build_scripts/qt/compilers/gcc.pri
@@ -1,25 +1,6 @@
 GCC_VERSION = $$system("g++ -dumpversion")
-greaterThan(GCC_VERSION, 6) {
-    message('g++' version is above 6. Using regular g++.)
-}
-else {
-    message('g++' version is not above 6. Attempting to use highest specific version.)
-    system(g++-7 --version) {
-        message('g++-7' found.)
-        QMAKE_CXX = g++-7
-        QMAKE_LINK = g++-7
-    }
-    system(g++-8 --version) {
-        message('g++-8' found.)
-        QMAKE_CXX = g++-8
-        QMAKE_LINK = g++-8
-    }
-    system(g++-9 --version) {
-        message('g++-9' found.)
-        QMAKE_CXX = g++-9
-        QMAKE_LINK = g++-9
-    }
-}
+QMAKE_CXX = g++
+QMAKE_LINK = g++
 
 system("$$QMAKE_CXX --version")
 
